pkg/runtime/wfengine/workflowstate.go

```diff
@@ -0,0 +1,232 @@
+	for i := len(events); i < removedCount; i++ {
+		req.Operations = append(req.Operations, actors.TransactionalOperation{
+			Operation: actors.Delete,
+			Request:   actors.TransactionalDelete{Key: getMultiEntryKeyName(keyPrefix, i)},
+		})
+	}
+	return nil
+}
+
+func LoadWorkflowState(ctx context.Context, actorRuntime actors.Actors, actorID string) (*workflowState, error) {
+	loadStartTime := time.Now()
+	loadedRecords := 0
+
+	req := actors.GetStateRequest{
+		ActorType: WorkflowActorType,
+		ActorID:   actorID,
+		Key:       metadataKey,
+	}
+	res, err := actorRuntime.GetState(ctx, &req)
+	loadedRecords++
+	if err != nil {
+		return nil, fmt.Errorf("failed to load workflow metadata: %w", err)
+	}
+	if len(res.Data) == 0 {
+		// no state found
+		return nil, nil
+	}
+	var metadata workflowStateMetadata
+	if err = json.Unmarshal(res.Data, &metadata); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal workflow metadata: %w", err)
+	}
+	state := NewWorkflowState(metadata.Generation)
+	// CONSIDER: Do some of these loads in parallel
+	for i := 0; i < metadata.InboxLength; i++ {
+		req.Key = getMultiEntryKeyName(inboxKeyPrefix, i)
+		res, err = actorRuntime.GetState(ctx, &req)
+		loadedRecords++
+		if err != nil {
+			return nil, fmt.Errorf("failed to load workflow inbox state key '%s': %w", req.Key, err)
+		}
+		var historyBytes []byte
+		if err = json.Unmarshal(res.Data, &historyBytes); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal JSON from inbox state key entry: %w", err)
+		}
+		var e *backend.HistoryEvent
+		e, err = backend.UnmarshalHistoryEvent(historyBytes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to unmarshal history event from inbox state key entry: %w", err)
+		}
+		state.Inbox = append(state.Inbox, e)
+	}
+	for i := 0; i < metadata.HistoryLength; i++ {
+		req.Key = getMultiEntryKeyName(historyKeyPrefix, i)
+		res, err = actorRuntime.GetState(ctx, &req)
+		loadedRecords++
+		if err != nil {
+			return nil, fmt.Errorf("failed to load workflow inbox state key '%s': %w", req.Key, err)
+		}
+		var historyBytes []byte
+		if err = json.Unmarshal(res.Data, &historyBytes); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal JSON from inbox state key entry: %w", err)
+		}
+		var e *backend.HistoryEvent
+		e, err = backend.UnmarshalHistoryEvent(historyBytes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to unmarshal history event from inbox state key entry: %w", err)
+		}
+
+		state.History = append(state.History, e)
+	}
+
+	req.Key = customStatusKey
+	res, err = actorRuntime.GetState(ctx, &req)
+	loadedRecords++
+	if err != nil {
+		return nil, fmt.Errorf("failed to load workflow custom status key '%s': %w", req.Key, err)
+	}
+	if err = json.Unmarshal(res.Data, &state.CustomStatus); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal JSON from inbox state key entry: %w", err)
+	}
+
+	wfLogger.Infof("%s: loaded %d state records in %v", actorID, loadedRecords, time.Since(loadStartTime))
```

Comment:
I like that we are tracking these and would love if we had metrics emitted for workflow operations; I don't think this PR is the place because it's somewhat tangential to this (I would like to make metric instruments easier to create and work with in general beforehand anyway) and can be done separately, but I think it's good to continue at least collecting and logging the important data so we can emit them. :+1: